From 14da01d4d5cc9771144f39ba92b327182b0c03c6 Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Fri, 15 Feb 2013 16:33:33 -0800
Subject: [PATCH 18/30] mesa: Statically allocate glthread command buffer in
 the batch struct.

This avoids an extra pointer dereference in the marshalling functions,
which, with the instruction count doing in the low 30s, could actually
matter for main-thread performance.
---
 src/mesa/main/glthread.c | 5 ++---
 src/mesa/main/glthread.h | 8 ++++----
 2 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/src/mesa/main/glthread.c b/src/mesa/main/glthread.c
index 01ca78c275..3a7c753fac 100644
--- a/src/mesa/main/glthread.c
+++ b/src/mesa/main/glthread.c
@@ -45,10 +45,10 @@ glthread_allocate_batch(struct gl_context *ctx)
    struct glthread_state *glthread = ctx->GLThread;
 
    /* TODO: handle memory allocation failure. */
-   glthread->batch = calloc(1, sizeof(*glthread->batch));
+   glthread->batch = malloc(sizeof(*glthread->batch));
    if (!glthread->batch)
       return;
-   glthread->batch->buffer = malloc(MARSHAL_MAX_CMD_SIZE);
+   memset(glthread->batch, 0, offsetof(struct glthread_batch, buffer));
 }
 
 static void
@@ -64,7 +64,6 @@ glthread_unmarshal_batch(struct gl_context *ctx, struct glthread_batch *batch)
 
    assert(pos == batch->used);
 
-   free(batch->buffer);
    free(batch);
 }
 
diff --git a/src/mesa/main/glthread.h b/src/mesa/main/glthread.h
index c38fef32fa..98ae11509a 100644
--- a/src/mesa/main/glthread.h
+++ b/src/mesa/main/glthread.h
@@ -94,14 +94,14 @@ struct glthread_batch
    struct glthread_batch *next;
 
    /**
-    * Points to the first command in the batch.
+    * Amount of data used by batch commands, in bytes.
     */
-   uint8_t *buffer;
+   size_t used;
 
    /**
-    * Amount of data used by batch commands, in bytes.
+    * Data contained in the command buffer.
     */
-   size_t used;
+   uint8_t buffer[MARSHAL_MAX_CMD_SIZE];
 };
 
 void _mesa_glthread_init(struct gl_context *ctx);
-- 
2.11.0

